enhancements/machine-api/cluster-autoscaler-integration.md

```diff
@@ -0,0 +1,449 @@
+  // Construct a NodeInfo that looks like what a Node from this NodeGroup would look like
+  // Eg capacity, labels and taints
+  ...
+}
+
+func (ng *nodegroup) Exist() bool {
+  // Not supporting dynamic NodeGroup provisioning
+	return true
+}
+
+func (ng *nodegroup) Create() (cloudprovider.NodeGroup, error) {
+  // Not supporting dynamic NodeGroup provisioning
+	return nil, cloudprovider.ErrAlreadyExist
+}
+
+func (ng *nodegroup) Delete() error {
+  // Not supporting dynamic NodeGroup provisioning
+	return cloudprovider.ErrNotImplemented
+}
+
+func (ng *nodegroup) Autoprovisioned() bool {
+  // Not supporting dynamic NodeGroup provisioning
+	return false
+}
+
+```
+
+#### Scalable Resource
+
+A `scalableResource` will be the lowest level interface that will represent either a MachineSet or MachineDeployment.
+It will be used by the `nodegroup` to retrieve information about, or manipulate the underlying MachineSet/MachineDeployment.
+
+```Go
+// scalableResource is a resource that can be scaled up and down by
+// adjusting its replica count field.
+type scalableResource interface {
+	// Id returns an unique identifier of the resource
+	ID() string
+
+	// MaxSize returns maximum size of the resource
+	MaxSize() int
+
+	// MinSize returns minimum size of the resource
+	MinSize() int
+
+	// Name returns the name of the resource
+	Name() string
+
+	// Namespace returns the namespace the resource is in
+	Namespace() string
+
+	// Nodes returns a list of all machines that already have or should become nodes that belong to this
+	// resource
+	Nodes() ([]string, error)
+
+	// SetSize() sets the replica count of the resource
+	SetSize(nreplicas int32) error
+
+	// Replicas returns the current replica count of the resource
+	Replicas() (int32, error)
+
+	// MarkMachineForDeletion marks machine for deletion
+	MarkMachineForDeletion(machine *Machine) error
+
+	// UnmarkMachineForDeletion unmarks machine for deletion
+	UnmarkMachineForDeletion(machine *Machine) error
+
+  // The following methods will be used to create a template Node for scaling from Zero
+	Labels() map[string]string
+	Taints() []apiv1.Taint
+	CanScaleFromZero() bool
+	InstanceCPUCapacity() (resource.Quantity, error)
+	InstanceMemoryCapacity() (resource.Quantity, error)
+	InstanceGPUCapacity() (resource.Quantity, error)
+	InstanceMaxPodsCapacity() (resource.Quantity, error)
+}
+```
+
+##### Marking Machines for deletion
+
+To ensure that the MachineSet or MachineDeployment scales down the desired Node, when asked to scale down,
+the `scalableResource` will apply `"machine.openshift.io/cluster-api-delete-machine"` as a label to the Machine that
+has been flagged for termination. It is expected that the MachineSet/MachineDeployment controller should terminate
+this instance before any other when it is scaled down.
+
+#### Agnostic API
+
+To support multiple versions of the Cluster API simultaneously, we will implement an internal API which is expected to
```

Comment:
may be " For the provider to support multiple machine API implementations e.g machine.openshift.io, SAP Gardener, cluster.x-k8s.io..."